lib/Parse/ParseBridging.cpp

```diff
@@ -0,0 +1,180 @@
+
+
+#include "swift/Parse/ParseBridging.h"
+#include "swift/Bridging/ASTGen.h"
+#include "swift/Parse/Parser.h"
+
+using namespace swift;
+
+BridgedASTContext BridgedLegacyParser::getASTContext() const {
+  return unbridged().Context;
+}
+
+BridgedDiagnosticEngine BridgedLegacyParser::getDiagnosticEngine() const {
+  return &unbridged().Diags;
+}
+
+BridgedDeclContext BridgedLegacyParser::getCurrDeclContext() const {
+  return unbridged().CurDeclContext;
+}
+
+BridgedStringRef BridgedLegacyParser::getSourceBuffer() const {
+  // FIXME: Should we consider partial parsing?
+  StringRef text =
+      handle->SourceMgr.getEntireTextForBuffer(*handle->SF.getBufferID());
+  return {text.data(), text.size()};
+}
+
+// MARK: Functions called from SwiftParser.
+
+BridgedExpr BridgedLegacyParser::parseExpr(BridgedSourceLoc loc,
+                                           BridgedDeclContext DC,
+                                           bool isExprBasic) const {
+  auto &P = unbridged();
+  auto PP = P.getParserPosition(loc.unbridged(), loc.unbridged());
+  P.CurDeclContext = DC.unbridged();
+  P.restoreParserPosition(PP);
+  ParserResult<Expr> result =
+      P.parseExprImpl(diag::expected_expr, isExprBasic, /*fromASTGen=*/true);
+  return result.getPtrOrNull();
```

Comment:
We somehow also need to propagate `hasError` and `hasCodeCompletionToken`. Is that something you want to do in this PR or in a follow-up?